README.md
# mpyq

mpyq is a Python library for reading MPQ (MoPaQ) archives used in many of
Blizzard's games. It was originally developed for data mining Starcraft II
replay files.

In addition to being a library, mpyq also has a command line interface that
exposes some of the library's core functionality such as extracting archives.

At this early stage in development only files compressed with DEFLATE or bzip2
are uncompressed. This means that this library can not be used to extract most
big game asset archives that Blizzard's games use. More compression formats
will be supported in the future.

Also, as mpyq is so far pure Python code, it might be unfeasible to try to
extract very large MPQ archives, even if all the compression methods used
inside the archive were supported.

Note that listing files inside an archive does not require full extraction.
You can safely take a peek inside any MPQ archive with this library.

## Installation

A stable version of mpyq is available from PyPI and can be installed with
either `easy_install` or `pip`.

    $ easy_install mpyq
    $ pip install mpyq

mpyq can be installed manually with the included setup.py script.

    $ python setup.py install

Running this command will install mpyq both as a library and a stand-alone
script that can be run from anywhere, provided that you have added Python's
bin directory to your PATH environment variable.

Alternative way to install mpyq is to clone this git repository somewhere on
your filesystem and then either adjust your PYTHONPATH environment variable to
point to the directory that contains the repository or create a symbolic link
to your Python's site-packages directory pointing at the repository.

Note that the command line interface part of mpyq uses the argparse module,
which was included into Python's standard library in version 2.7. If you
didn't install mpyq from PyPI and you wish to use the command line interface
part with Python 2.6, you should install argparse from PyPI manually.

## Usage

### As a library

    >>> from mpyq import MPQArchive
    >>> archive = MPQArchive('game.SC2Replay')

Now you have a MPQArchive object of the file you opened. One common thing
to do now is to extract the files from the archive.

    >>> files = archive.extract()

This will extract and return the archive's contents in memory. Be advised
that you might not want to do this with multi-gigabyte MPQ files from
World of Warcraft, for example.

Files inside the archive can be also extracted and written to disk.

    >>> archive.extract_to_disk()

If you want to skip reading the (listfile) inside the archive, you can do
so by passing `listfile=False` to the constructor.

    >>> archive = MPQArchive('bad_listfile.SC2Replay', listfile=False)

This might be required if the (listfile) is encrypted or has been tampered
with. Note that you can't list files or extract the whole archive if you do
this -- you need to know in advance which files you want to read.

    >>> archive.read('replay.details')
    '\x05\x1c\x00\x04\x01\x00\x04\x05...'

For more information, consult `help(mpyq)` in your Python console.

### From the command line

    usage: mpyq.py [-h] [-I] [-H] [-b] [-s] [-t] [-x] file

    mpyq reads and extracts MPQ archives.

    positional arguments:
      file                 path to the archive

    optional arguments:
      -h, --help           show this help message and exit
      -I, --headers        print header information from the archive
      -H, --hash-table     print hash table
      -b, --block-table    print block table
      -s, --skip-listfile  skip reading (listfile)
      -t, --list-files     list files inside the archive
      -x, --extract        extract files from the archive

You can extract all the files inside the archive with `-x/--extract`.

    $ mpyq -x game.SC2Replay

This will create a directory called 'game' with the files inside.

You can print the header information from a given archive with `-I/--headers`.

    $ mpyq -I game.SC2Replay
    MPQ archive header
    ------------------
    magic                          'MPQ\x1a'
    header_size                    44
    arhive_size                    19801
    format_version                 1
    sector_size_shift              3
    hash_table_offset              19385
    block_table_offset             19641
    hash_table_entries             16
    block_table_entries            10
    extended_block_table_offset    0
    hash_table_offset_high         0
    block_table_offset_high        0
    offset                         1024

    MPQ user data header
    --------------------
    magic                          'MPQ\x1b'
    user_data_size                 512
    mpq_header_offset              1024
    user_data_header_size          60
    content                        '\x05\x08\x00\x02,StarCraft II replay\x1b
                                    11\x02\x05\x0c\x00\t\x02\x02\t\x00\x04\t
                                    (\x06\t\x00\x08\t\xc8\xfa\x01\n\t\xc8\xf
                                    a\x01\x04\t\x04\x06\t\xa2\x99\x01'

You can display the archive's hash table with `-H/--hash-table`.

    $ mpyq -H game.SC2Replay
    MPQ archive hash table
    ----------------------
     Hash A   Hash B  Locl Plat BlockIdx
    D38437CB 07DFEAEC 0000 0000 00000009
    AAC2A54B F4762B95 0000 0000 00000002
    FFFFFFFF FFFFFFFF FFFF FFFF FFFFFFFF
    FFFFFFFF FFFFFFFF FFFF FFFF FFFFFFFF
    FFFFFFFF FFFFFFFF FFFF FFFF FFFFFFFF
    C9E5B770 3B18F6B6 0000 0000 00000005
    343C087B 278E3682 0000 0000 00000004
    3B2B1EA0 B72EF057 0000 0000 00000006
    5A7E8BDC FF253F5C 0000 0000 00000001
    FD657910 4E9B98A7 0000 0000 00000008
    D383C29C EF402E92 0000 0000 00000000
    FFFFFFFF FFFFFFFF FFFF FFFF FFFFFFFF
    FFFFFFFF FFFFFFFF FFFF FFFF FFFFFFFF
    FFFFFFFF FFFFFFFF FFFF FFFF FFFFFFFF
    1DA8B0CF A2CEFF28 0000 0000 00000007
    31952289 6A5FFAA3 0000 0000 00000003

You can display the archive's block table with `-b/--block-table`.

    $ mpyq -b game.SC2Replay
    MPQ archive block table
    -----------------------
     Offset  ArchSize RealSize  Flags
    0000002C      443      443 81000200
    000001E7      609     1082 81000200
    00000448    16077    42859 81000200
    00004315       94       94 81000200
    00004373       96       96 81000200
    000043D3      591      765 81000200
    00004622      802     1444 81000200
    00004944      247      580 81000200
    00004A3B      120      164 81000200
    00004AB3      262      288 81000200

You can list all files inside the archive with `-t/--list-files`.

    $ mpyq -t game.SC2Replay
    Files
    -----
    replay.attributes.events            580 bytes
    replay.details                      443 bytes
    replay.game.events                42859 bytes
    replay.initData                    1082 bytes
    replay.load.info                     96 bytes
    replay.message.events                94 bytes
    replay.smartcam.events             1444 bytes
    replay.sync.events                  765 bytes

You can skip reading the listfile with `-s/--skip-listfile`. This might be
necessary if the listfile is encrypted or corrupted. Note that you cannot
list files or extract the whole archive without the listfile.

## References

The following two documents were used as references for the MPQ format:

 * [http://www.zezula.net/en/mpq/mpqformat.html](http://www.zezula.net/en/mpq/mpqformat.html)
 * [http://wiki.devklog.net/index.php?title=The_MoPaQ_Archive_Format](http://wiki.devklog.net/index.php?title=The_MoPaQ_Archive_Format)


## Copyright

Copyright 2010-2014 Aku Kotkavuo. See LICENSE for details.


CHANGELOG.md
# CHANGE LOG

### 0.2.5 - Releaed 2014-03-17

 * Fixed Python 3 incompatibilities

### 0.2.4 - Released 2014-01-05

 * Fixed bug in multi-sector block decompression

### 0.2.3 - Released 2013-07-31

 * Fixed division bug with Python3 compatibility

### 0.2.2 - Released 2013-04-08

 * Added Python3 support

### 0.2.1 - Released 2013-02-25

 * Now properly writes empty archive files to disk
 * Improved handling of invalid archive files, now raises ValueError

### 0.2.0 - Released 2011-09-26

 * Allow for MPQ file extraction to current directory
 * Returns None when reading empty files from the archive

### 0.1.11 - Released 2011-09-20

 * Automatically install argparse dependency during install

### 0.1.10 - Released 2011-09-20

 * Fixes bug when using force_decompress option on a multi-sector block

### 0.1.9 - Released 2011-09-18

 * Introduces force_decompress option. This flexibility allows for reading
   files from some otherwise corrupted MPQ files.
 * Fixes print_headers for archives without user_data_header
 * Improves formatting of print_file command

### 0.1.8 - Released 2011-03-28

 * Adds support for opening any file-like object (objects implementing .read())
 * Returns None when attempting to read missing or non-existant file from
   archive

### 0.1.7 - Released 2011-03-08

 * Adds support for opening archives from opened file objects
 
### 0.1.6 - Released 2011-03-03

 * Fixes decompression bug for multi-sector blocks

### 0.1.5 - Released 2011-03-01

 * Adds option to skip/ignore the list file. Allows reading when listfile has
   been encrypted or tampered with when you already know what you are looking
   for.

### 0.1.4 - Released 2011-02-28

 * Raises NotImplementedError when attempting to read encrypted files from
   archives

### 0.1.3 - Released 2011-02-26

 * Adds print_hash_table and print_block_table commands

### 0.1.2 - Released 2011-02-24

 * Fixed UnboundLocalError in read_file

### 0.1.1 - Released 2011-01-05

 * Fixed installation layout
 * Added minimal regression test

### 0.1.0 - Released 2010-09-05

 * Released on PyPI for the first time

### Initial Release - 2010-05-15



test/test_mpqarchive.py
#!/usr/bin/env python
# coding: utf-8

import os
import sys
import unittest
from mpyq import MPQArchive

TEST_DIR = os.path.realpath(os.path.dirname(__file__)) + '/'

class TestMPQArchive(unittest.TestCase):

    def setUp(self):
        self.archive = MPQArchive(TEST_DIR + 'test.SC2Replay')

    def test_init_with_file(self):
        self.archive = MPQArchive(open(TEST_DIR + 'test.SC2Replay', 'rb'))

    def test_header(self):
        self.assertEqual(self.archive.header['magic'], b'MPQ\x1a')
        self.assertEqual(self.archive.header['header_size'], 44)
        self.assertEqual(self.archive.header['archive_size'], 205044)
        self.assertEqual(self.archive.header['format_version'], 1)
        self.assertEqual(self.archive.header['sector_size_shift'], 3)
        self.assertEqual(self.archive.header['hash_table_offset'], 204628)
        self.assertEqual(self.archive.header['block_table_offset'], 204884)
        self.assertEqual(self.archive.header['hash_table_entries'], 16)
        self.assertEqual(self.archive.header['block_table_entries'], 10)
        self.assertEqual(self.archive.header['extended_block_table_offset'], 0)
        self.assertEqual(self.archive.header['hash_table_offset_high'], 0)
        self.assertEqual(self.archive.header['block_table_offset_high'], 0)
        self.assertEqual(self.archive.header['offset'], 1024)

    def test_files(self):
        self.assertEqual(self.archive.files, [b'replay.attributes.events',
                                              b'replay.details',
                                              b'replay.game.events',
                                              b'replay.initData',
                                              b'replay.load.info',
                                              b'replay.message.events',
                                              b'replay.smartcam.events',
                                              b'replay.sync.events'])

    def test_print_hash_table(self):
        self.archive.print_hash_table()
        self.assertEqual(sys.stdout.getvalue(),
                         "MPQ archive hash table\n"
                         "----------------------\n"
                         " Hash A   Hash B  Locl Plat BlockIdx\n"
                         "D38437CB 07DFEAEC 0000 0000 00000009\n"
                         "AAC2A54B F4762B95 0000 0000 00000002\n"
                         "FFFFFFFF FFFFFFFF FFFF FFFF FFFFFFFF\n"
                         "FFFFFFFF FFFFFFFF FFFF FFFF FFFFFFFF\n"
                         "FFFFFFFF FFFFFFFF FFFF FFFF FFFFFFFF\n"
                         "C9E5B770 3B18F6B6 0000 0000 00000005\n"
                         "343C087B 278E3682 0000 0000 00000004\n"
                         "3B2B1EA0 B72EF057 0000 0000 00000006\n"
                         "5A7E8BDC FF253F5C 0000 0000 00000001\n"
                         "FD657910 4E9B98A7 0000 0000 00000008\n"
                         "D383C29C EF402E92 0000 0000 00000000\n"
                         "FFFFFFFF FFFFFFFF FFFF FFFF FFFFFFFF\n"
                         "FFFFFFFF FFFFFFFF FFFF FFFF FFFFFFFF\n"
                         "FFFFFFFF FFFFFFFF FFFF FFFF FFFFFFFF\n"
                         "1DA8B0CF A2CEFF28 0000 0000 00000007\n"
                         "31952289 6A5FFAA3 0000 0000 00000003\n"
                         "\n")

    def test_print_block_table(self):
        self.archive.print_block_table()
        self.assertEqual(sys.stdout.getvalue(),
                         "MPQ archive block table\n"
                         "-----------------------\n"
                         " Offset  ArchSize RealSize  Flags\n"
                         "0000002C      727      890 81000200\n"
                         "00000303      801     1257 81000200\n"
                         "00000624   194096   479869 81000200\n"
                         "0002FC54      226      334 81000200\n"
                         "0002FD36       97       97 81000200\n"
                         "0002FD97     1323     1970 81000200\n"
                         "000302C2     6407    12431 81000200\n"
                         "00031BC9      533     2400 81000200\n"
                         "00031DDE      120      164 81000200\n"
                         "00031E56      254      288 81000200\n"
                         "\n")


mpyq.py
#!/usr/bin/env python
# coding: utf-8

"""
mpyq is a Python library for reading MPQ (MoPaQ) archives.
"""

from __future__ import print_function

import bz2
import os
import struct
import zlib
from collections import namedtuple
from io import BytesIO


__author__ = "Aku Kotkavuo"
__version__ = "0.2.5"


MPQ_FILE_IMPLODE        = 0x00000100
MPQ_FILE_COMPRESS       = 0x00000200
MPQ_FILE_ENCRYPTED      = 0x00010000
MPQ_FILE_FIX_KEY        = 0x00020000
MPQ_FILE_SINGLE_UNIT    = 0x01000000
MPQ_FILE_DELETE_MARKER  = 0x02000000
MPQ_FILE_SECTOR_CRC     = 0x04000000
MPQ_FILE_EXISTS         = 0x80000000

MPQFileHeader = namedtuple('MPQFileHeader',
    '''
    magic
    header_size
    archive_size
    format_version
    sector_size_shift
    hash_table_offset
    block_table_offset
    hash_table_entries
    block_table_entries
    '''
)
MPQFileHeader.struct_format = '<4s2I2H4I'

MPQFileHeaderExt = namedtuple('MPQFileHeaderExt',
    '''
    extended_block_table_offset
    hash_table_offset_high
    block_table_offset_high
    '''
)
MPQFileHeaderExt.struct_format = 'q2h'

MPQUserDataHeader = namedtuple('MPQUserDataHeader',
    '''
    magic
    user_data_size
    mpq_header_offset
    user_data_header_size
    '''
)
MPQUserDataHeader.struct_format = '<4s3I'

MPQHashTableEntry = namedtuple('MPQHashTableEntry',
    '''
    hash_a
    hash_b
    locale
    platform
    block_table_index
    '''
)
MPQHashTableEntry.struct_format = '2I2HI'

MPQBlockTableEntry = namedtuple('MPQBlockTableEntry',
    '''
    offset
    archived_size
    size
    flags
    '''
)
MPQBlockTableEntry.struct_format = '4I'


class MPQArchive(object):

    def __init__(self, filename, listfile=True):
        """Create a MPQArchive object.

        You can skip reading the listfile if you pass listfile=False
        to the constructor. The 'files' attribute will be unavailable
        if you do this.
        """
        if hasattr(filename, 'read'):
            self.file = filename
        else:
            self.file = open(filename, 'rb')
        self.header = self.read_header()
        self.hash_table = self.read_table('hash')
        self.block_table = self.read_table('block')
        if listfile:
            self.files = self.read_file('(listfile)').splitlines()
        else:
            self.files = None

    def read_header(self):
        """Read the header of a MPQ archive."""

        def read_mpq_header(offset=None):
            if offset:
                self.file.seek(offset)
            data = self.file.read(32)
            header = MPQFileHeader._make(
                struct.unpack(MPQFileHeader.struct_format, data))
            header = header._asdict()
            if header['format_version'] == 1:
                data = self.file.read(12)
                extended_header = MPQFileHeaderExt._make(
                    struct.unpack(MPQFileHeaderExt.struct_format, data))
                header.update(extended_header._asdict())
            return header

        def read_mpq_user_data_header():
            data = self.file.read(16)
            header = MPQUserDataHeader._make(
                struct.unpack(MPQUserDataHeader.struct_format, data))
            header = header._asdict()
            header['content'] = self.file.read(header['user_data_header_size'])
            return header

        magic = self.file.read(4)
        self.file.seek(0)

        if magic == b'MPQ\x1a':
            header = read_mpq_header()
            header['offset'] = 0
        elif magic == b'MPQ\x1b':
            user_data_header = read_mpq_user_data_header()
            header = read_mpq_header(user_data_header['mpq_header_offset'])
            header['offset'] = user_data_header['mpq_header_offset']
            header['user_data_header'] = user_data_header
        else:
            raise ValueError("Invalid file header.")

        return header

    def read_table(self, table_type):
        """Read either the hash or block table of a MPQ archive."""

        if table_type == 'hash':
            entry_class = MPQHashTableEntry
        elif table_type == 'block':
            entry_class = MPQBlockTableEntry
        else:
            raise ValueError("Invalid table type.")

        table_offset = self.header['%s_table_offset' % table_type]
        table_entries = self.header['%s_table_entries' % table_type]
        key = self._hash('(%s table)' % table_type, 'TABLE')

        self.file.seek(table_offset + self.header['offset'])
        data = self.file.read(table_entries * 16)
        data = self._decrypt(data, key)

        def unpack_entry(position):
            entry_data = data[position*16:position*16+16]
            return entry_class._make(
                struct.unpack(entry_class.struct_format, entry_data))

        return [unpack_entry(i) for i in range(table_entries)]

    def get_hash_table_entry(self, filename):
        """Get the hash table entry corresponding to a given filename."""
        hash_a = self._hash(filename, 'HASH_A')
        hash_b = self._hash(filename, 'HASH_B')
        for entry in self.hash_table:
            if (entry.hash_a == hash_a and entry.hash_b == hash_b):
                return entry

    def read_file(self, filename, force_decompress=False):
        """Read a file from the MPQ archive."""

        def decompress(data):
            """Read the compression type and decompress file data."""
            compression_type = ord(data[0:1])
            if compression_type == 0:
                return data
            elif compression_type == 2:
                return zlib.decompress(data[1:], 15)
            elif compression_type == 16:
                return bz2.decompress(data[1:])
            else:
                raise RuntimeError("Unsupported compression type.")

        hash_entry = self.get_hash_table_entry(filename)
        if hash_entry is None:
            return None
        block_entry = self.block_table[hash_entry.block_table_index]

        # Read the block.
        if block_entry.flags & MPQ_FILE_EXISTS:
            if block_entry.archived_size == 0:
                return None

            offset = block_entry.offset + self.header['offset']
            self.file.seek(offset)
            file_data = self.file.read(block_entry.archived_size)

            if block_entry.flags & MPQ_FILE_ENCRYPTED:
                raise NotImplementedError("Encryption is not supported yet.")

            if not block_entry.flags & MPQ_FILE_SINGLE_UNIT:
                # File consist of many sectors. They all need to be
                # decompressed separately and united.
                sector_size = 512 << self.header['sector_size_shift']
                sectors = block_entry.size // sector_size + 1
                if block_entry.flags & MPQ_FILE_SECTOR_CRC:
                    crc = True
                    sectors += 1
                else:
                    crc = False
                positions = struct.unpack('<%dI' % (sectors + 1),
                                          file_data[:4*(sectors+1)])
                result = BytesIO()
                sector_bytes_left = block_entry.size
                for i in range(len(positions) - (2 if crc else 1)):
                    sector = file_data[positions[i]:positions[i+1]]
                    if (block_entry.flags & MPQ_FILE_COMPRESS and
                        (force_decompress or sector_bytes_left > len(sector))):
                        sector = decompress(sector)

                    sector_bytes_left -= len(sector)
                    result.write(sector)
                file_data = result.getvalue()
            else:
                # Single unit files only need to be decompressed, but
                # compression only happens when at least one byte is gained.
                if (block_entry.flags & MPQ_FILE_COMPRESS and
                    (force_decompress or block_entry.size > block_entry.archived_size)):
                    file_data = decompress(file_data)

            return file_data

    def extract(self):
        """Extract all the files inside the MPQ archive in memory."""
        if self.files:
            return dict((f, self.read_file(f)) for f in self.files)
        else:
            raise RuntimeError("Can't extract whole archive without listfile.")

    def extract_to_disk(self):
        """Extract all files and write them to disk."""
        archive_name, extension = os.path.splitext(os.path.basename(self.file.name))
        if not os.path.isdir(os.path.join(os.getcwd(), archive_name)):
            os.mkdir(archive_name)
        os.chdir(archive_name)
        for filename, data in self.extract().items():
            f = open(filename, 'wb')
            f.write(data or b'')
            f.close()

    def extract_files(self, *filenames):
        """Extract given files from the archive to disk."""
        for filename in filenames:
            data = self.read_file(filename)
            f = open(filename, 'wb')
            f.write(data or b'')
            f.close()

    def print_headers(self):
        print("MPQ archive header")
        print("------------------")
        for key, value in self.header.items():
            if key == "user_data_header":
                continue
            print("{0:30} {1!r}".format(key, value))
        if self.header.get('user_data_header'):
            print()
            print("MPQ user data header")
            print("--------------------")
            for key, value in self.header['user_data_header'].items():
                print("{0:30} {1!r}".format(key, value))
        print()

    def print_hash_table(self):
        print("MPQ archive hash table")
        print("----------------------")
        print(" Hash A   Hash B  Locl Plat BlockIdx")
        for entry in self.hash_table:
            print('{0:0>8X} {1:0>8X} {2:0>4X} {3:0>4X} {4:0>8X}'.format(*entry))
        print()

    def print_block_table(self):
        print("MPQ archive block table")
        print("-----------------------")
        print(" Offset  ArchSize RealSize  Flags")
        for entry in self.block_table:
            print('{0:0>8X} {1:>8} {2:>8} {3:>8X}'.format(*entry))
        print()

    def print_files(self):
        if self.files:
            print("Files")
            print("-----")
            width = max(len(name) for name in self.files) + 2
            for filename in self.files:
                hash_entry = self.get_hash_table_entry(filename)
                block_entry = self.block_table[hash_entry.block_table_index]
                print("{0:{width}} {1:>8} bytes".format(filename.decode(),
                                                        block_entry.size,
                                                        width=width))

    def _hash(self, string, hash_type):
        """Hash a string using MPQ's hash function."""
        hash_types = {
            'TABLE_OFFSET': 0,
            'HASH_A': 1,
            'HASH_B': 2,
            'TABLE': 3
        }
        seed1 = 0x7FED7FED
        seed2 = 0xEEEEEEEE

        for ch in string.upper():
            if not isinstance(ch, int): ch = ord(ch)
            value = self.encryption_table[(hash_types[hash_type] << 8) + ch]
            seed1 = (value ^ (seed1 + seed2)) & 0xFFFFFFFF
            seed2 = ch + seed1 + seed2 + (seed2 << 5) + 3 & 0xFFFFFFFF

        return seed1

    def _decrypt(self, data, key):
        """Decrypt hash or block table or a sector."""
        seed1 = key
        seed2 = 0xEEEEEEEE
        result = BytesIO()

        for i in range(len(data) // 4):
            seed2 += self.encryption_table[0x400 + (seed1 & 0xFF)]
            seed2 &= 0xFFFFFFFF
            value = struct.unpack("<I", data[i*4:i*4+4])[0]
            value = (value ^ (seed1 + seed2)) & 0xFFFFFFFF

            seed1 = ((~seed1 << 0x15) + 0x11111111) | (seed1 >> 0x0B)
            seed1 &= 0xFFFFFFFF
            seed2 = value + seed2 + (seed2 << 5) + 3 & 0xFFFFFFFF

            result.write(struct.pack("<I", value))

        return result.getvalue()

    def _prepare_encryption_table():
        """Prepare encryption table for MPQ hash function."""
        seed = 0x00100001
        crypt_table = {}

        for i in range(256):
            index = i
            for j in range(5):
                seed = (seed * 125 + 3) % 0x2AAAAB
                temp1 = (seed & 0xFFFF) << 0x10

                seed = (seed * 125 + 3) % 0x2AAAAB
                temp2 = (seed & 0xFFFF)

                crypt_table[index] = (temp1 | temp2)

                index += 0x100

        return crypt_table

    encryption_table = _prepare_encryption_table()


def main():
    import argparse
    description = "mpyq reads and extracts MPQ archives."
    parser = argparse.ArgumentParser(description=description)
    parser.add_argument("file", action="store", help="path to the archive")
    parser.add_argument("-I", "--headers", action="store_true", dest="headers",
                        help="print header information from the archive")
    parser.add_argument("-H", "--hash-table", action="store_true",
                        dest="hash_table", help="print hash table"),
    parser.add_argument("-b", "--block-table", action="store_true",
                        dest="block_table", help="print block table"),
    parser.add_argument("-s", "--skip-listfile", action="store_true",
                        dest="skip_listfile", help="skip reading (listfile)"),
    parser.add_argument("-t", "--list-files", action="store_true", dest="list",
                        help="list files inside the archive")
    parser.add_argument("-x", "--extract", action="store_true", dest="extract",
                        help="extract files from the archive")
    args = parser.parse_args()
    if args.file:
        if not args.skip_listfile:
            archive = MPQArchive(args.file)
        else:
            archive = MPQArchive(args.file, listfile=False)
        if args.headers:
            archive.print_headers()
        if args.hash_table:
            archive.print_hash_table()
        if args.block_table:
            archive.print_block_table()
        if args.list:
            archive.print_files()
        if args.extract:
            archive.extract_to_disk()


if __name__ == '__main__':
    main()
